### ISSUE TITLE
Synchronize file output between snakemake and nextflow versions

### ISSUE THREAD

**Issue by photocyte** [transXpress/transXpress-snakemake#13]
Here is what the nextflow version is currently outputting in a transXpress_results folder in the working directory of the pipeline execution (using rnaSPAdes as an example):
```
0226_Chattonella_subsalsa_annotated.fasta
0226_Chattonella_subsalsa_annotated.pep
0226_Chattonella_subsalsa_rnaSPAdes.fasta
0226_Chattonella_subsalsa_rnaSPAdes.fasta.transdecoder_dir/
0226_Chattonella_subsalsa_rnaSPAdes.fasta.transdecoder_dir.__checkpoints_longorfs/
0226_Chattonella_subsalsa_rnaSPAdes.fasta.transdecoder.pep
assembly_seq-dependent_checksums.txt
blastp_annotations.tsv
blastx_annotations.tsv
deeploc_annotations.txt
fastqc_results/
kallisto.gene.TMM.EXPR.matrix
kallisto.gene.TPM.not_cross_norm
kallisto.isoform.TMM.EXPR.matrix
kallisto.isoform.TPM.not_cross_norm
pfam_annotations.txt
pfam_dom_annotations.txt
pipeliner.782.cmds
rfam_annotations.txt
rnaSPAdes_renamed.fasta.gene_trans_map
tmhmm_annotations.tsv
transcriptome_exN50.plot.pdf
transcriptome_stats.txt
transcriptome_TPM_blast.csv
```

I'd actually like to give every file the assemblyPrefix + assembler ID, as if I run the same pipeline twice with trinity / rnaspades, I don't want files to be overwritten. E.g. `0226_Chattonella_subsalsa_annotated.pep` isn't acceptable as it isn't unique between assemblers. Also possible that with some rearranging that the nextflow pipeline could execute trinity/rnaspades simultaneously, so important things don't get overwritten. Alternatively, could publish assembly-specific files to their own subdirectory (e.g. 0226_Chattonella_subsalsa_rnaSPAdes) to prevent overwriting.

**Comment by tomas-pluskal** [comment]
I think we should not say that the pipeline can be executed in the same folder twice with different parameters - it somewhat defies the philosophy of snakemake.

The output of the pipeline (including all the intermediate files) ultimately depends on the input (samples.txt) and the configuration (config.yaml). If any of those two are changed, the whole pipeline has to be executed from scratch. If you want to produce multiple transcriptomes with different settings in the same working folder, it means we have to place ALL output (including all the intermediate files) into a separate folder that will be specific to that particular execution. But how to identify that execution? The assembly prefix will never be complex enough to capture all the settings in the config.yaml.

I think a better strategy is to say that the pipeline can be executed once in one folder with particular input. It has several advantages:
- snakemake will always know what the intermediate files are and can re-execute rules as needed without worrying about mixing intermediate files from different executions
- if something is changed in the samples.txt or config.yaml, snakemake will notice and reexecute the whole pipeline
- the final transcriptome and the state of the folder after the execution of the pipeline will always correspond to the input sample.txt and config.yaml, which is useful for retracing the steps (and writing Materials and Methods)

If the user wants to execute the pipeline with different parameters, it is trivial to copy the Snakefile, samples.txt and config.yaml to a different folder and do a separate execution in that folder with different parameters.

**Comment by tomas-pluskal** [comment]
if you think it is useful to have both trinity and rnaspades transcriptomes, we could add an option like ```assembler: "both"``` to the config.yaml. But in any case, it means the transdecoder, expression analysis, and annotation steps will be executed separately for both transcriptomes, so I feel it is better to run the pipeline twice in different folders, once with ```assembler: "trinity"``` and once with ```assembler: "rnaspades"```.

The only step that would be shared for the two runs would be the trimmomatic step, but that doesn't seem like a major advantage. On the other hand, if the pipelines are executed in two different folders, it increases the chance that at least one would finish without error :)

**Comment by photocyte** [comment]
I agree it is simpler to run things twice in different folders.  Let's plan for that.  But I'll still plan to try and fix the naming so `
0226_Chattonella_subsalsa_annotated.pep` -> `
0226_Chattonella_subsalsa_annotated_rnaSPAdes.pep`

Not sure if I love the .pep vs .fasta naming scheme for distinguising transcripts versus peptides (since both are .fasta files)

**Comment by tomas-pluskal** [comment]
I agree with using .fasta instead of .pep files. How about this naming?

````
prefix_assembler_transcripts.fasta
prefix_assembler_transcripts_annotated.fasta
prefix_assembler_proteins.fasta
prefix_assembler_proteins_annotated.fasta
````

For example

````
20190226_Chattonella_subsalsa_rnaSPAdes_transcripts.fasta
20190226_Chattonella_subsalsa_rnaSPAdes_transcripts_annotated.fasta
20190226_Chattonella_subsalsa_rnaSPAdes_proteins.fasta
20190226_Chattonella_subsalsa_rnaSPAdes_proteins_annotated.fasta
````

If we insist on adding a date, let's use the full scale yyyymmdd, scientific projects always span multiple years. I am not sure if we should insist on species - some people might be doing weird stuff (meta-transcriptomics, unknown species, etc), or may want to index their samples by something else than species name. In my view, it would be better to specify the prefix format as a recommendation, not as a hard rule.

**Comment by photocyte** [comment]
I agree with the naming you suggest.  
Shouldn't be too hard to implement, will try to do it in Nextflow version soon.

I think the YYYY is overkill, especially if the prefix is being added onto the FASTA record ids (which is my current intent, and nextflow version behavior).  My interest in using MMDD was just to distinguish different assemblies done on the same species within a single lab.  "v1,v2" would suffice as well, but MMDD has some nice properties compared to "v1,v2" (e.g. increasing over short timescales, easy understood/agreed upon value), and although collisions can happen, since assemblies for the most part take longer than a day, it seems a relatively safe thing to add.  Adding YYYY would prevent collisions across years, but that is a bit rare, and wouldn't solve the main issue of collisions for two assemblies done on the same day.  

The species key need not actually correspond to an accepted species, currently there is no check that it actually corresponds to a valid species, so it can be any type of metadata people would like to input.  

I'd draw a distinction between the proposed "hard rule" being, "you must include some type of metadata by default", versus, "you must include a valid species name as your mandatory default metadata".  I think the former rule is good practice, and minimally invasive if presented with a switch to turn that behavior off, while the latter rule is not so valuable in the absence of downstream processing which uses that genus/species information, and indeed, isn't even sensible in certain use cases.

Can also have different levels of renaming, with different switches.  E.g. switches for annotate with date, annotate with metadata, annotate with assembler.  Could have them all on by default, with the metadata value left blank, but the default behavior would be to reject a blank metadata value.  That way, if someone loves the FASTA annotation, but doesn't love the assembler annotation/date/metadata, they can turn all the assembly/record renaming off, but keep the annotation.

**Comment by photocyte** [comment]
Current transXpress-nextflow output with trinity
```
0301_Photinus_pyralis_Trinity.proteins_annotated.fasta
0301_Photinus_pyralis_Trinity.transcripts_annotated.fasta
0301_Photinus_pyralis_Trinity.transcripts.fasta
0301_Photinus_pyralis_Trinity.transcripts.fasta.transdecoder_dir/
0301_Photinus_pyralis_Trinity.transcripts.fasta.transdecoder_dir.__checkpoints_longorfs/
0301_Photinus_pyralis_Trinity.transcripts.fasta.transdecoder.pep
assembly_seq-dependent_checksums.txt
blastp_annotations.tsv
blastx_annotations.tsv
deeploc_annotations.txt
fastqc_results/
kallisto.gene.TMM.EXPR.matrix
kallisto.gene.TPM.not_cross_norm
kallisto.isoform.TMM.EXPR.matrix
kallisto.isoform.TPM.not_cross_norm
pfam_annotations.txt
pfam_dom_annotations.txt
pipeliner.12482.cmds
rfam_annotations.txt
tmhmm_annotations.tsv
transcriptome_exN50.plot.pdf
transcriptome.fasta
transcriptome.fasta.gene_trans_map
transcriptome_stats.txt
transcriptome_TPM_blast.csv
trinity_out_dir/
Trinity_renamed.fasta.gene_trans_map
```

Current transXpress-nextflow output with rnaspades
```
0301_Photinus_pyralis_rnaSPAdes.proteins_annotated.fasta
0301_Photinus_pyralis_rnaSPAdes.transcripts_annotated.fasta
0301_Photinus_pyralis_rnaSPAdes.transcripts.fasta
0301_Photinus_pyralis_rnaSPAdes.transcripts.fasta.transdecoder_dir/
0301_Photinus_pyralis_rnaSPAdes.transcripts.fasta.transdecoder_dir.__checkpoints_longorfs/
0301_Photinus_pyralis_rnaSPAdes.transcripts.fasta.transdecoder.pep
assembly_seq-dependent_checksums.txt
blastp_annotations.tsv
blastx_annotations.tsv
deeploc_annotations.txt
fastqc_results/
kallisto.gene.TMM.EXPR.matrix
kallisto.gene.TPM.not_cross_norm
kallisto.isoform.TMM.EXPR.matrix
kallisto.isoform.TPM.not_cross_norm
pfam_annotations.txt
pfam_dom_annotations.txt
pipeliner.14951.cmds
rfam_annotations.txt
rnaSPAdes_renamed.fasta.gene_trans_map
tmhmm_annotations.tsv
transcriptome_exN50.plot.pdf
transcriptome_stats.txt
transcriptome_TPM_blast.csv
```

**Comment by tomas-pluskal** [comment]
Re: MMDD discussion
Looking back at the many different kava transcriptomes I produced over the last 4 years, if they were named like
````
0301_Piper_methysticum.transcripts.fasta
0503_Piper_methysticum.transcripts.fasta
1201_Piper_methysticum.transcripts.fasta
1220_Piper_methysticum.transcripts.fasta
````
the date is not very informative or useful. Perhaps we should just stick to v1, v2, v3..? That will guarantee some logical ordering, at least, and it is short and unobtrusive.

If we define the preferred prefix as
```<species>_<assembler>_v<version>```
the resulting files would be named
```
Chattonella_subsalsa_rnaSPAdes_v1_transcripts.fasta
Chattonella_subsalsa_rnaSPAdes_v1_transcripts_annotated.fasta
Chattonella_subsalsa_rnaSPAdes_v1_proteins.fasta
Chattonella_subsalsa_rnaSPAdes_v1_proteins_annotated.fasta
```
What do you think?

**Comment by photocyte** [comment]
The date isn't intended to be very informative, as I don't see any benefit on knowing the exact day when an assembly was produced, but instead just serve as a unique identifier.  The sorting by time is a nice characteristic, but maybe not so beneficial, as there is no guarantee that a later assembly is necessarily any better than an earlier one, so I don't see it as essential.  I'd expect doing 4 assemblies over 4 years is a less common case, not sure if it should drive the solution.

For the v1,v2,v3,etc versioning, then one has to keep the state of the current version.  Which, is non trivial across different users / directory structures / filesystems / servers etc.  If you have a good solution to that, then maybe it is worth doing.  I don't think having it be a user configurable parameter is a good solution (then you just end up with multiple competing v1s or v2s that are hard to distinguish).

I think there are multiple goals for a naming prefix, which no solution solves entirely:
1) Unique without user input (default parameters)
2) Resistant to collisions
3) Stateless 
4) Informative
5) Short

I would say
MMDD ~= YYMMDD > YYYYMMDD > v1,v2,v3 versioning

Also P.S. the slight change from the format of your suggestion:
```
0301_Photinus_pyralis_rnaSPAdes_transcripts.fasta
```

to

```
0301_Photinus_pyralis_rnaSPAdes.transcripts.fasta
```

Which I find easier to read, and also distinguishes the invariant / formatting description from the metadata description.

**Comment by tomas-pluskal** [comment]
Ok, we can do it that way, and see if any future users have further comments/suggestions/objections.

So, shall we add 4 configuration options to both pipelines?

1. prefix_add_date: yes/no, default yes
2. prefix_add_metadata: yes/no, default yes
3. prefix_metadata: default empty, recommended Genus_species, will produce an error if empty and prefix_add_metadata is yes
4. prefix_add_assembler: yes/no, default yes

I'm fine with using the dot instead of underscore for .transcripts and .proteins filenames.

By the way, I would suggest to keep the assembler names all lowercase (trinity and rnaspades), both for the configuration options and for the final final filenames. Keeping everything lowercase eliminates the ambiguity whether the settings and filenames are case-sensitive or not.

**Comment by photocyte** [comment]
One could even imagine making the date formatting string a configurable option.  In any event, I'll plan to have it set to YYMMDD by default, with a link to the respective Python or Groovy documentation on the date formatting.

For the assembler names, I think it looks nicer and is more easily readable to keep the canonical capitalization.  E.g. `Trinity` `rnaSPAdes`.  I agree can run into case-senstivity issues, but there should only be a few places these strings are compared & a few .toUpper() should handle that.

**Comment by tomas-pluskal** [comment]
If the user specifies the assembler as "rNaSpAdEs", would you name the final file as "0301_Photinus_pyralis_rnaSPAdes.transcripts.fasta" or "0301_Photinus_pyralis_rNaSpAdEs.transcripts.fasta"?

For the prefix file, it's fine for me, but in that case I'd call the file something like "assembly_prefix.txt", because people might want to put a lot more stuff in there than just the species name.

**Comment by photocyte** [comment]
My intent is the former case.  In the nextflow version for example, currently the param.assembler string (obtained from the config file) is used in a string regex comparison to direct the program execution, however the comparison has an embedded toLowerCase() for the param.assembler string, so it is case insensitive.  rNaSpAdEs would execute appropriately. But for the actual assembler string that is used to name the output files (e.g. rnaSPAdes), rather than using a global variable, the string is hardcoded as an output from the assembly process node, which the downstream nodes then use.  This is robust to running two different assemblers at the same time & getting the same naming correct, but the global variable approach does seem simpler if only expecting to run one assembler at a time.
